crates/typed-store/src/sally/mod.rs

```diff
@@ -0,0 +1,588 @@
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.multi_insert(key_val_pairs),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.multi_insert(key_val_pairs),
+        }
+    }
+    fn multi_remove<J>(&self, keys: impl IntoIterator<Item = J>) -> Result<(), Self::Error>
+    where
+        J: Borrow<K>,
+    {
+        self.batch().delete_batch(self, keys)?.write()
+    }
+    fn try_catch_up_with_primary(&self) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(db_map.try_catch_up_with_primary()?),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(test_db.try_catch_up_with_primary()?),
+        }
+    }
+}
+
+impl<J, K, U, V> TryExtend<(J, U)> for SallyColumn<K, V>
+where
+    J: Borrow<K> + std::clone::Clone,
+    U: Borrow<V> + std::clone::Clone,
+    K: Serialize,
+    V: Serialize,
+{
+    type Error = TypedStoreError;
+
+    fn try_extend<T>(&mut self, iter: &mut T) -> Result<(), Self::Error>
+    where
+        T: Iterator<Item = (J, U)>,
+    {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend(iter),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend(iter),
+        }
+    }
+    fn try_extend_from_slice(&mut self, slice: &[(J, U)]) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend_from_slice(slice),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend_from_slice(slice),
+        }
+    }
+}
+
+/// A Sally write batch provides a mutable struct which holds a collection of db mutation operations and
+/// applies them atomically to the db.
+pub enum SallyWriteBatch {
+    // Write batch for RocksDB backend when `fallback_to_db` is set as true
+    RocksDB(DBBatch),
+    // Write batch for btree map based backend
+    TestDB(TestDBWriteBatch),
+}
+
+impl SallyWriteBatch {
```

Comment:
Couple of questions about batch:
- Should we associate a batch with an ID and ensure that only one batch with this ID commits (this may require mapping the ID to some condition in the DB).
- Should we allow reads from the batch that reflect what was written?
- Should we allow transactional semantics, where we can do reads in a batch, and specify that the read should still be the same as a condition for the batch to commit?

Reply:
> * Should we allow reads from the batch that reflect what was written?

This is definitely possible, and a good idea. I can add it to our batch api

> * Should we allow transactional semantics, where we can do reads in a batch, and specify that the read should still be the same as a condition for the batch to commit?

This is something in the works. I have an almost finished PR which exposes optimistic transactions in DBMap (our rocksdb library) which sally is going to start using to support transactions once that PR is landed. So, we will have something like:
```
let txn1 = db.transaction();
let v1 = txn1.get_for_update(b"k1").unwrap();
txn1.put(b"k2", b"v2").unwrap();
txn1.commit(); // this fails if another txn txn2 wrote "k1" or "k2" 
```
>* Should we associate a batch with an ID and ensure that only one batch with this ID commits (this may require mapping the ID to some condition in the DB).
It sounds like this could be achieved with transactions as well, where we will do a read on batch ID to lock it, so something like:
```
let txn1 = db.transaction();
// read batch id key 100 (id for this batch)
let val = txn1.get_for_update(b"id=100"); 
// if a different tx already wrote a batch with this id, fail immediately
if val.is_some() {
  // another txn already wrote a batch with id = 100
  txn1.rollback();
} else {
  txn1.put(b"k1", b"v1");
  txn1.put(b"k2, b"v2");
  txn1.put(b"k3", b"v3");
}
txn1.put(b"id=100", "xyz");
txn1.commit(); // If another transaction wrote batch id=100 since it was read in this transaction, this commit will fail
```

Reply:
Awesome!